lib/Dialect/TTNN/Transforms/Passes.cpp

```diff
@@ -3,20 +3,105 @@
-#include "mlir/Support/LogicalResult.h"
-#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
-#include "ttmlir/Conversion/TTIRToTTNN/TTIRToTTNN.h"
-#include "ttmlir/Dialect/TT/IR/TT.h"
-#include "ttmlir/Dialect/TT/IR/TTOpsTypes.h"
-#include "ttmlir/Dialect/TTIR/IR/TTIR.h"
-#include "ttmlir/Dialect/TTIR/IR/TTIROps.h"
-#include "ttmlir/Dialect/TTIR/Transforms/Passes.h"
 #include "ttmlir/Dialect/TTNN/IR/TTNNOpsTypes.h"
 
-namespace mlir::tt::ttnn {} // namespace mlir::tt::ttnn
+namespace mlir::tt::ttnn {
+#define GEN_PASS_DEF_TTNNDEALLOCATE
+#include "ttmlir/Dialect/TTNN/Transforms/Passes.h.inc"
+
+class TTNNDeallocate : public impl::TTNNDeallocateBase<TTNNDeallocate> {
+
+public:
+  using impl::TTNNDeallocateBase<TTNNDeallocate>::TTNNDeallocateBase;
+
+  Operation *getLastValueUsageOp(const LivenessBlockInfo *livenessInfo,
+                                 Value value) {
+    Operation *startOp = livenessInfo->getStartOperation(value);
+    Operation *endOp = livenessInfo->getEndOperation(value, startOp);
+    auto *opOperandIter =
+        llvm::find_if(endOp->getOpOperands(), [&](OpOperand &opOperand) {
+          return opOperand.is(value);
+        });
+
+    // Keep going until we find the last usage of the tensor.
+    // For DPS stop if operand is not dps init one.
+    // For non DPS skip through as it might modify the tensor in place.
+    //
+    while (opOperandIter != endOp->getOpOperands().end() &&
+           ((isa<DestinationStyleOpInterface>(endOp) &&
+             cast<DestinationStyleOpInterface>(endOp).isDpsInit(
+                 &(*opOperandIter))) ||
+            (!isa<DestinationStyleOpInterface>(endOp) &&
+             endOp->getNumResults() > 0))) {
+      assert(endOp->getResults().size() == 1);
+      OpResult result = endOp->getResult(0);
+      endOp = livenessInfo->getEndOperation(result, endOp);
+      opOperandIter =
+          llvm::find_if(endOp->getOpOperands(), [&](OpOperand &opOperand) {
+            return opOperand.is(result);
+          });
+    }
+
+    return endOp;
+  }
+
+  void runOnOperation() final {
+    ModuleOp module = getOperation();
+    IRRewriter rewriter(&getContext());
+
+    module->walk([&](func::FuncOp func) {
+      assert(func.getBody().hasOneBlock());
+      Liveness liveness(func.getOperation());
+      const LivenessBlockInfo *livenessInfo =
+          liveness.getLiveness(&func.getBody().front());
+
+      // Handle non DPS ops which do not store function result and are used to
+      // allocate tensors. DPS ops are handled via ttnn::EmptyOp.
+      //
+      func->walk([&](Operation *op) {
+        if (isa<DestinationStyleOpInterface>(op)) {
+          return;
+        }
+
+        // Skip ops which do not have results.
+        //
+        if (op->getNumResults() == 0) {
+          return;
+        }
+
+        // Check if result is ranked tensor type.
+        //
+        if (!isa<RankedTensorType>(op->getResult(0).getType())) {
+          return;
+        }
```

Comment:
We should probably handle cases where there's multiple results and check if any of them is an RTT. Maxpool is going to have 2 outputs in forward graph given there's a training graph to follow.

Reply:
Correct, one will be the "actual" result, the other will be metainfo on which indices were chosen for each pooling window (not sure what the format will be).

Reply:
Multiple result Ops are supported now.